<a href="https://colab.research.google.com/github/YaPineiro/Fin420/blob/main/YP__week_0_session_2_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Packages


Say you create a set of functions in Excel. To share these functions you would have to send your spreadsheet to someone else.  The other person then opens your (entire) spreadsheet and copies and pastes the wanted formulas.  This is a hackish way to interact.

To share functions in Python, you put them into a package and submit the package to a package manager.  If anyone wants to use your package (functions/data structures) they install and load your package.  In this way, you can easily and safely share your code with millions of other people.  Alternatively, you can use easily use code written by others.

## Implications for Recreating Others' Work

In Excel you are constantly recreating functions you and others have already done.  This is due to the **inefficiency of sharing code between spreadsheets.**

In Python, once we are someone else creates a function, it can be easily shared.  Thus, we generally will not have to create our own functions, but rather we can use those created by others (note this makes the code documentation extra important).  This **frees us to spend our time understanding the implications of our calculations, and how it informs our financial decision**.  


<b>Python makes you a better financial analyst because you can focus on where you add value (making financial decisions), rather than on perpetually reimplementing calclations.</b>






## Python Packages in Colab

Most of the packages/modules we will use are preinstalled in Colab, and so we only need to load them.  Some packages we will have to install however.





See [here for a more complete description of the difference between a python package and module](https://stackoverflow.com/questions/7948494/whats-the-difference-between-a-python-module-and-a-python-package).

## Preinstalled Packages

Examples of a couple important preinstalled packages that we will use are `numpy` and `pandas`.  

Say we need to generate random numbers from a standard normal distribution.  We can load numpy's random module and then:


To load preinstalled packages we only need to:

In [1]:
import numpy as np
import pandas as pd

We can now use the packages:

In [2]:
np.sinh(4)

27.28991719712775

In [3]:
np.random.randn(3,3)

array([[-1.79773951,  0.70822446,  0.16609215],
       [ 0.62076052,  1.07866944, -0.49201295],
       [-2.2266123 , -0.95517844, -0.06737229]])

In [4]:
pd.DataFrame([[1, 2],
              [3, 4]])

0  1
0  1  2
1  3  4

## Importing Packages

Above we wrote `import numpy as np`.  The `as np` portion simply allows us to a shorter way to refer to numpy and is unnecessary. If we leave it out we refer to functions in the package with `numpy.` instead of `np.`.

We refer to functions (such as `sinh`) provided by numpy with `np.sinh`.  This ensures there is no confusion with any other functions called `sinh` in other packages.

### Methods of Import

In others' Python code you may see:

`import numpy`

or

`from numpy import sinh`

While there are differences, it doesn't really matter which you use. For explanations see [this answer on Stackoverflow](https://stackoverflow.com/questions/710551/use-import-module-or-from-module-import) and [here in the Python documentation](https://docs.python.org/3/tutorial/modules.html#packages).

In [5]:
from numpy import random

In [6]:
random.randn(3,3)

array([[ 0.57901965, -0.12506327, -0.28719767],
       [ 0.4197095 ,  1.91843809,  0.07108273],
       [-0.16017834, -0.35211837, -0.45118467]])

### Find Module Functions

In [7]:
dir(np)

['ALLOW_THREADS',
 'BUFSIZE',
 'CLIP',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'WRAP',
 '_CopyMode',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__deprecated_attrs__',
 '__dir__',
 '__doc__',
 '__expired_functions__',
 '__file__',
 '__former_attrs__',
 '__future_scalars__',
 '__getattr__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_add_newdoc_ufunc',
 '_builtins',
 '_core',
 '_distributor_init',


In [8]:
np.array([[1, 2, 3.0], [2, 3, 4]]).shape

(2, 3)

## Installing Packages in Colab

However some Python packages we will use are not preinstalled.  We therefore must installed them prior to loading.  For example, if we try to load the bankfind package we get an error: `No module named 'bankfind'`

In [9]:
import bankfind

ModuleNotFoundError: No module named 'bankfind'

We need to first install the `bankfind` package by executing the following BASH command:

In [10]:
!pip install bankfind

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.1/41.1 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.2/47.2 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.8/61.8 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 kB 10.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 128.0/128.0 kB 10.5 MB/s eta 0:00:00
  Attempting uninstall: chardet
    Found existing installation: chardet 5.2.0
    Uninstalling chardet-5.2.0:
      Successfully uninstalled chardet-5.2.0
  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.3.0
    Uninstalling urllib3-2.3.0:
      Successfully uninstalled urllib3-2.3.0
  Attempting uninstall: idna
    Found existing installation: idna 3.10
    Uninstalling idna-3.10:
      Successfully uninstalled idna-3.10
  Attempting uninstall: requests
    Found existing installation: reque

In [11]:
import bankfind as bf

In [12]:
vermont_json = bf.get_institutions(filters="STNAME:Vermont")
vermont_json

{'meta': {'total': 63,
  'parameters': {'filters': 'STNAME:Vermont',
   'fields': 'ACTIVE,ADDRESS,ASSET,BKCLASS,CB,CBSA,CBSA_DIV,CBSA_DIV_FLG,CBSA_DIV_NO,CBSA_METRO,CBSA_METRO_FLG,CBSA_METRO_NAME,CBSA_MICRO_FLG,CBSA_NO,CERT,CERTCONS,CFPBEFFDTE,CFPBENDDTE,CFPBFLAG,CHANGEC1,CHANGEC2,CHANGEC3,CHANGEC4,CHANGEC5,CHANGEC6,CHANGEC7,CHANGEC8,CHANGEC9,CHANGEC10,CHANGEC11,CHANGEC12,CHANGEC13,CHANGEC14,CHANGEC15,CHARTER,CHRTAGNT,CITY,CITYHCR,CLCODE,CMSA_NO,CMSA,CONSERVE,COUNTY,CSA,CSA_NO,CSA_FLG,DATEUPDT,DENOVO,DEP,DEPDOM,DOCKET,EFFDATE,ENDEFYMD,EQ,ESTYMD,FDICDBS,FDICREGN,FDICSUPV,FED,FED_RSSD,FEDCHRTR,FLDOFF,FORM31,HCTMULT,IBA,INACTIVE,INSAGNT1,INSAGNT2,INSBIF,INSCOML,INSDATE,INSDIF,INSFDIC,INSSAIF,INSSAVE,INSTAG,INSTCRCD,LAW_SASSER_FLG,MSA,MSA_NO,MUTUAL,NAME,NAMEHCR,NETINC,NETINCQ,NEWCERT,OAKAR,OCCDIST,OFFDOM,OFFFOR,OFFICES,OFFOA,OTSDIST,OTSREGNM,PARCERT,PROCDATE,QBPRCOML,REGAGNT,REGAGENT2,REPDTE,RISDATE,ROA,ROAPTX,ROAPTXQ,ROAQ,ROE,ROEQ,RSSDHCR,RUNDATE,SASSER,SPECGRP,SPECGRPN,STALP,STALPHCR,STC

In [13]:
vermont_json['data']

[{'ZIP': '05403',
  'SASSER': 0,
  'CHRTAGNT': 'STATE',
  'CONSERVE': 'N',
  'REGAGENT2': '',
  'STNAME': 'Vermont',
  'ROAQ': -0.74,
  'INSDATE': '08/02/2022',
  'TE06N528': '',
  'TE06N529': '',
  'OFFOA': 0,
  'FDICDBS': 2,
  'NAMEHCR': '',
  'OCCDIST': '',
  'CMSA': '',
  'DEPDOM': 230108,
  'CBSA_METRO_FLG': '1',
  'TE10N528': '',
  'NETINC': -1581,
  'CBSA_DIV_NO': '',
  'MUTUAL': '0',
  'OFFFOR': 0,
  'INSSAVE': 0,
  'CHARTER': '0',
  'RSSDHCR': '',
  'TE04N528': '',
  'TE04N529': '',
  'CERT': 59298,
  'STALP': 'VT',
  'SPECGRP': 4,
  'CFPBENDDTE': '9999-12-31',
  'TE09N528': '',
  'IBA': 0,
  'INSBIF': 0,
  'INSFDIC': 1,
  'ENDEFYMD': '12/31/9999',
  'MSA': '',
  'TE02N528': '',
  'CB': '1',
  'TE02N529': '',
  'TE07N528': '',
  'FDICSUPV': 'New York',
  'FED': '01',
  'REGAGNT': 'FDIC',
  'NEWCERT': 0,
  'ASSET': 257839,
  'CBSA_MICRO_FLG': '0',
  'OFFICES': 1,
  'STCNTY': '50007',
  'CSA_FLG': '1',
  'CITY': 'South Burlington',
  'CLCODE': '21',
  'INACTIVE': 0,
  'STALPHCR'

## A Note on JSON

These data are in a format known as JSON, which stands for Javascript Object Notation.  It is a convenient format for storing and transerring data (you can read more here: https://www.json.org/json-en.html).  Pandas known how to handle this type of data, however, and so we can convert it from JSON to a `DataFrame` easily with `json_normalize`.

In [14]:
banks_vermont = pd.json_normalize(vermont_json['data'])
banks_vermont[banks_vermont["NAMEHCR"].notna()]

ZIP  SASSER CHRTAGNT CONSERVE REGAGENT2   STNAME   ROAQ     INSDATE  \
0   05403       0    STATE        N            Vermont  -0.74  08/02/2022   
1   05760       0      OCC        N            Vermont   1.28  01/01/1934   
2   05478       0    STATE        N            Vermont   1.38  04/02/1934   
3   05301       0    STATE        N            Vermont   1.22  08/09/1989   
4   05201       0      OCC        N            Vermont   0.79  08/09/1989   
5   05081       0    STATE        N            Vermont   0.42  03/31/1934   
6   05753       0      OCC        N            Vermont   0.87  01/01/1934   
7   05055       0      OCC        N            Vermont   0.44  05/22/1991   
8   05855       0      OCC        N            Vermont   1.20  01/01/1934   
9   05819       0    STATE        N            Vermont   1.07  03/31/1934   
10  05663       0    STATE        N            Vermont   0.29  03/31/1934   
11  05661       0    STATE        N            Vermont   0.42  03/31/1934   
15  05401       0      OCC        N            Vermont   1.09  03/31/1934   
16  05701       0    STATE        N            Vermont   1.05  03/31/1934   
19  05641       0    STATE        N            Vermont   1.85  03/31/1934   
20  05101       0    STATE        N            Vermont   0.41  01/01/1934   
21  05301       0      OCC        N            Vermont   1.62  03/30/1934   
22  05401       0      OTS        N            Vermont   1.14  08/09/1989   
25  05656       0    STATE        N            Vermont   1.00  04/02/1934   
26  05851       0    STATE        N            Vermont  -2.83  04/02/1934   
28  05701       0    STATE        N            Vermont   0.00  12/31/9999   
29  05478       0      OCC        N            Vermont   1.95  03/31/1934   
30  05819       0    STATE        N            Vermont   0.73  03/31/1934   
32  05001       0    STATE        N            Vermont -23.27  04/02/1934   
33  05672       0    STATE        N            Vermont   3.10  02/05/1977   
34  05340       0    STATE        N            Vermont -20.92  02/14/1977   
35  05001       0    STATE        N            Vermont -14.97  03/23/1987   
39  05156       0    STATE        N            Vermont   0.54  08/09/1989   
40  05091       0    STATE        N            Vermont  -0.42  11/14/1989   
41  05033       0      OCC        N            Vermont   0.15  01/29/1993   
42  05156       0      OCC        N            Vermont   0.21  01/01/1934   
46  05033       0      OCC        N            Vermont  -5.40  01/01/1934   
47  05733       0      OCC        N            Vermont   1.14  01/01/1934   
48  05301       0      OCC        N            Vermont   0.94  01/01/1934   
49  05401       0      OCC        N            Vermont   1.55  01/01/1934   
50  05403       0    STATE        N            Vermont   0.80  01/01/1934   
51  05828       0      OCC        N            Vermont  -1.82  01/01/1934   
54  05255       0      OCC        N            Vermont   1.37  01/01/1934   
57  05060       0      OCC        N            Vermont  -0.43  01/01/1934   
59  05033       0      OCC        N            Vermont   0.14  01/01/1934   
62  05091       0      OCC        N            Vermont   1.31  01/01/1934   

   TE06N528 TE06N529  ...  TE08N528  NETINCQ      ESTYMD FEDCHRTR TRUST  \
0                     ...             -441.0  08/02/2022        0     0   
1                     ...              267.0  01/01/1863        1     0   
2                     ...             1282.0  11/12/1866        0     1   
3                     ...              867.0  01/01/1912        0     1   
4                     ...             1138.0  01/01/1917        1     0   
5                     ...              251.0  01/01/1892        0     0   
6                     ...             1205.0  11/09/1831        1     1   
7                     ...             1009.0  05/22/1991        1     1   
8                     ...             3415.0  01/15/1851        1     1   
9                    

Which returns a list of banks in Vermont.  Note the use of `.notna()`.  We'll unpack the line `banks_vermont[banks_vermont["NAMEHCR"].notna()]` in following sessions.  However, here it is worth knowing that it returns the `banks_vermont` DataFrame with any row where the value of `NAMEHCR` is removed.

# Finding Python Packages

To search for python packages you can install via pip, you can use the PyPi website: [https://pypi.org/](https://pypi.org/).  Note, you used to be able (and may be able to agin in the near future) use `pip search` at the command line.  For example we could search for any finance packages with:

In [15]:
!pip search finance

ERROR: XMLRPC request failed [code: -32500]
RuntimeError: PyPI no longer supports 'pip search' (or XML-RPC search). Please use https://pypi.org/search (via a browser) instead. See https://warehouse.pypa.io/api-reference/xml-rpc.html#deprecated-methods for more information.


However this functionality has been temporarily/permanently removed.

# Using Packages

The package's pip page, will usually link to the source code repository.  This is most likely a GitHub or GitLab repo.  The repo will contain a README.md file which will explain how to use the package (or link to the documentation housed elsewhere).  

For example, here is [the repo with the README.md for the bankfind package](https://github.com/dpguthrie/bankfind).  The README.md file is automatically displayed below the code.  It is a fairly simple package so the README overview is short.  The README does link to more detailed documentation, as well as the documentation for the underlying data from the FDIC.  

## Security

It is possible that there is malicious code in Python packages hosted on pip.  The probability is low however, particularly if you are using well-known packages.  Here are a couple of pointers:

-  Read the source code.  It doesn't take as long as you think.
-  Don't run packages with administrator privileges.
-  Only install packages with Github or Gitlab repos that you can access (goes to point 1).

In general package security isn't much of an issue, and if you ever have concerns about a package run it in something like Colab.


# Exercise

*The following exercise may not work. The `cbpro` package doesn't work with python 3.10+ and the alternative `coinbasepro` package also won't load*

Find a package that wraps the coinbase-pro API ([this package may do the trick](https://github.com/danpaquin/coinbasepro-python)).  Use the package to get [a list of available currency pairs for trading](https://docs.pro.coinbase.com//#get-products).  All the code you will need to get the list of currency pairs is in the [GitHub README](https://github.com/danpaquin/coinbasepro-python).

# Alternative Exercise

Find a package that will allow you to download data from Yahoo Finance. Use the package to download stock data for ticker `MSFT`.

# Alternative Exercise Solution

In [3]:
!pip install yfinance

In [11]:
import yfinance as yfi

In [12]:
ticker = "MSFT"

In [16]:
data = yfi.download(ticker, start="2024-01-01", end="2024-12-31")

[*********************100%***********************]  1 of 1 completed


In [17]:
print(data.head())

Price            Close        High         Low        Open    Volume
Ticker            MSFT        MSFT        MSFT        MSFT      MSFT
Date                                                                
2024-01-02  368.117249  373.109913  364.047674  371.085046  25258600
2024-01-03  367.849274  370.489534  365.774790  366.271079  23083500
2024-01-04  365.208984  370.330688  364.444711  367.918732  20901500
2024-01-05  365.020416  369.298423  363.779694  366.231362  20987000
2024-01-08  371.908905  372.415129  366.271072  366.558897  23134000


# Coinbase Exercise Solution

In [16]:
!pip install cbpro

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.5/44.5 kB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.2/196.2 kB 15.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 44.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 584.6/584.6 kB 33.3 MB/s eta 0:00:00
  Created wheel for pymongo: filename=pymongo-3.5.1-cp311-cp311-linux_x86_64.whl size=376141 sha256=a087a5e69985bffaffc1d66e1dd140c3b8325f729241db199a2b89d886a7fc63
  Stored in directory: /root/.cache/pip/wheels/ed/ed/4d/71debf050a349e79742ff761bba17b37d7fbb905ba4a0eb58c
  Created wheel for websocket-client: filename=websocket_client-0.40.0-py2.py3-none-any.whl size=198285 sha256=bcd06e42ed98d1403038dfbc7010bbbd16ab131dfa39273d7393adde8ddcefc4
  Stored in directory: /root/.cache/pip/wheels/46/b8/c5/6e165f9cb05e74eac2d8898b59cff7637cfb1f577bdbe52c61
Successfully built pymongo websocket-cli

In [17]:
import cbpro

ImportError: cannot import name 'MutableMapping' from 'collections' (/usr/lib/python3.11/collections/__init__.py)

In [1]:
!pip install coinbasepro

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 5.1 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.13.0
    Uninstalling requests-2.13.0:
      Successfully uninstalled requests-2.13.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bankfind 0.0.1 requires requests==2.24.0, but you have requests 2.32.3 which is incompatible.
cbpro 1.1.4 requires requests==2.13.0, but you have requests 2.32.3 which is incompatible.
tensorflow 2.17.1 requires six>=1.12.0, but you have six 1.10.0 which is incompatible.


In [2]:
import coinbasepro